PDE docstrings

nicoguaro opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Closes #12370
Closes #16020

#### Brief description of what is fixed or changed

Update documentation for PDE utils.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

nicoguaro committed:
Add general solution for the PDEs to the docstrings

nicoguaro committed:
Add general solution for the PDEs to the docstrings

nicoguaro committed:
Correct typo in docstring

nicoguaro committed:
Correct typo in docstring

nicoguaro committed:
Remove trailing space in pde.py

nicoguaro committed:
Remove trailing space in pde.py

nicoguaro committed:
Remove trailing space in pde.py

nicoguaro committed:
Remove trailing spaces in pde.py

Remove trailing spaces at line 500

nicoguaro committed:
Merge branch 'pde_docstrings' of github.com:nicoguaro/sympy into pde_docstrings

nicoguaro committed:
Fix indentation in pde_linear_variable_coeff

nicoguaro committed:
Fix indentation in pde_linear_variable_coeff

nicoguaro committed:
Merge branch 'pde_docstrings' of github.com:nicoguaro/sympy into pde_docstrings

sylee957 commented on sympy/solvers/pde.py:
```diff
@@ -567,7 +572,17 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
+        \frac{- b \eta + c \xi}{b^2 + c^2} \right )
+        e^{\frac{a \xi}{b^2 + c^2}}\, d\xi\right)
+        e^{- \frac{a \xi}{b^2 + c^2}}
+        \right|_{\substack{ \eta=- b y + c x\\ \xi=b x + c y }}
```
I think your formula is not aligned with the equation

<table><tr><td><img src="https://user-images.githubusercontent.com/34944973/53186658-c1496700-3644-11e9-9b4b-7cd9200f99de.png" width="512px"></td></tr></table>


You would have to change `b, c` to `a, b` like below

<table><tr><td><img src="https://user-images.githubusercontent.com/34944973/53186848-3026c000-3645-11e9-8d16-4e10db721fa7.png" width="512px"></td></tr></table>

I also find that the example section had not aligned coefficients with the 
Viktor Grigoryan, “Partial Differential Equations” Math 124A - Fall 2010, pp.7.
Can you also edit the coefficients in example?

nicoguaro replied:
Done.

sylee957 replied:
<img src="https://user-images.githubusercontent.com/34944973/53244661-e4c8ec00-36ee-11e9-9351-6f1d3307444f.png" width="256px">

`e^{\frac{a \xi}{a^2 + b^2}}`  ->  `e^{\frac{c \xi}{a^2 + b^2}}`

nicoguaro replied:
Done.

sylee957 commented on sympy/solvers/pde.py:
```diff
@@ -662,26 +677,26 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
 
     Making the following substitutions reduces it to the linear ODE
 
     .. math:: a(\xi, \eta)\frac{du}{d\xi} + c(\xi, \eta)u - d(\xi, \eta) = 0
```
I think `d(xi, eta)` had come from nowhere

<img src="https://user-images.githubusercontent.com/34944973/53187399-6153c000-3646-11e9-82b7-f565d6cb9870.png" width="256px">

But after reading Viktor Grigoryan's reference, I think it is naively denoting `G(xi, eta)`,
So either this one can be changed to `G(xi, eta)`, or the equation above can be changed to `d(xi, eta)`.

And also, the example section should align with the new formula

nicoguaro replied:
Done.

sylee957 commented on sympy/solvers/pde.py:
```diff
@@ -662,26 +677,26 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
-    which can be solved using dsolve.
-
-    The general form of this PDE is::
+    which can be solved using ``dsolve``.
```
dsolve -> pdsolve

nicoguaro replied:
This one is correct, though. It says that after making the substitutions it can be solve using ``dsolve`` that is what is described in the book and the code is doing.

nicoguaro committed:
Update PDE docstrings

- Change coefficients in general form for pde_1st_linear_constant_coeff
- Change coefficients in general form for pde_1st_linear_variable_coeff

nicoguaro commented on sympy/solvers/pde.py:
```diff
@@ -662,26 +677,26 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
 
     .. math:: a(x, y) \frac{df(x, y)}{dx} + a(x, y) \frac{df(x, y)}{dy}
-                + c(x, y) f(x, y) - G(x, y)
+                + c(x, y) f(x, y) = G(x, y)
```
Done.

oscarbenjamin commented:
Doctests are failing because of unicode symbols. The doctests implicitly use unicode=False so you should do the same before pasting in the example:
```julia
In [15]: pprint(beta, use_unicode=True)                                                                                                        
β

In [16]: pprint(beta, use_unicode=False)                                                                                                       
beta
```

nicoguaro commented:
@oscarbenjamin, I will do those changes. Although, why it does not allows unicode?, even my name requires unicode to be written.

nicoguaro committed:
Remove unicode from docstrings

oscarbenjamin commented:
You can use unicode in general but the doctests use the ASCII printer.

Sympy uses all kinds of Symbols that don't render properly in many editors or in gmail etc. I think it's best to keep those out of the code where possible.

If you look at [https://github.com/sympy/sympy/pull/16032/commits/9bba75cf4914570a6a57309116c4a7619db9386d#diff-d9c12e4e549dfec9de253d529f20ba11R307](https://github.com/sympy/sympy/pull/16032/commits/9bba75cf4914570a6a57309116c4a7619db9386d#diff-d9c12e4e549dfec9de253d529f20ba11R307) you can see that Github is unable to render some symbols.

asmeurer commented:
That's not GitHub but the fonts on your computer. Font support for Unicode characters is definitely getting better, especially for common things like Greek characters. For things like the line drawing characters we use for parentheses it isn't so great everywhere, especially on devices like phones where the preinstalled fonts generally don't have full Unicode support for space concerns.

Not to mention it needs to be supported in the *monospace* font to look correct. See https://github.com/sympy/sympy/issues/15700. I just checked and `β` is not in the Courier font on my computer (the default monospace in Chrome and Firefox), meaning it would look broken in the rendered docs as long as #15700 remains unfixed.

oscarbenjamin commented:
Sorry I was being sloppy with my language there. Although Github itself is not at fault the end result is that I can't review the diff on Github in my browser. So I don't think we should have that sort of thing in the code unless necessary.

We can of course have names like Nicolás in the comments if needed (while Python 2.7 is supported that means the .py file needs a utf-8 declaration at the top) but we should stick to ASCII where possible.

I guess #15700 explains why many of the more complicated unicode printer examples I paste into Github end up misaligned.

nicoguaro committed:
Fix typo  ->

oscarbenjamin commented on sympy/solvers/pde.py:
```diff
@@ -567,7 +572,17 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
+    The general solution of the PDE is:
+
+    .. math::
+        f(x, y) = \left. \left[F(\eta) + \frac{1}{a^2 + b^2}
```
What is F here?

nicoguaro replied:
According to the reference is an arbitrary single-valued function.

oscarbenjamin replied:
Can you explain that in the docstring?

nicoguaro replied:
I added a comment about this in the docstrings.

oscarbenjamin reviewed (commented):
I left one comment. Otherwise this looks good.

smichr commented on sympy/solvers/pde.py:
```diff
@@ -567,7 +572,17 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
+        e^{- \frac{c \xi}{a^2 + b^2}}
+        \right|_{\substack{\eta=- a y + b x\\ \xi=a x + b y }}
+
+    and can be found in sympy with ``pdsolve``::
```
sympy->SymPy

smichr commented on sympy/solvers/pde.py:
```diff
@@ -496,7 +496,12 @@ def pde_1st_linear_constant_coeff_homogeneous(eq, func, order, match, solvefun):
+    .. math::
+        f(x, y) = F(- a y + b x ) e^{- \frac{c (a x + b y)}{a^2 + b^2}}
+
+    and can be found in sympy with ``pdsolve``::
```
sympy->SymPy

nicoguaro committed:
Fix typos in pde.py sympy >> SymPy

oscarbenjamin commented on sympy/solvers/pde.py:
```diff
@@ -662,39 +677,40 @@ def pde_1st_linear_constant_coeff(eq, func, order, match, solvefun):
 
-    .. math:: a(x, y) \frac{df(x, y)}{dx} + a(x, y) \frac{df(x, y)}{dy}
-                + c(x, y) f(x, y) - G(x, y)
+    .. math:: a(x, y) \frac{df(x, y)}{dx} + b(x, y) \frac{df(x, y)}{dy}
```
Can the `d`s be `\partial`?

nicoguaro replied:
Done.

nicoguaro committed:
Change differential for partial differentials in pde.py

oscarbenjamin commented:
Thanks very much for this @nicoguaro. This is a good improvement. Sorry for the delay in merging - I'm doing it now!
